loadtest/store/memstore/store.go

```diff
@@ -974,3 +986,113 @@ func (s *MemStore) SetServerVersion(version string) error {
 	s.serverVersion = version
 	return nil
 }
+
+// SetThread stores the given thread reponse.
+func (s *MemStore) SetThread(thread *model.ThreadResponse) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if thread == nil {
+		return errors.New("memstore: thread should not be nil")
+	}
+
+	// We get an element from the queue and check if we have it in the map and
+	// if it points to the same memory location. If so, we delete it since it means the queue is full.
+	// This is done to keep the data pointed by the map consistent with the data stored in the queue.
+	t := s.threadsQueue.Get().(*model.ThreadResponse)
+	if tt, ok := s.threads[t.PostId]; ok && tt == t {
+		delete(s.threads, t.PostId)
+	}
+	cloneThreadResponse(thread, t)
+	s.threads[thread.PostId] = t
+
+	return nil
+}
+
+// SetThreads stores the given thread response as a thread.
+func (s *MemStore) SetThreads(trs []*model.ThreadResponse) error {
+	if len(trs) == 0 {
+		mlog.Info("No threads to add to store")
+		return nil
+	}
+	for _, tr := range trs {
+		if err := s.SetThread(tr); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *MemStore) getThreads(unreadOnly bool) ([]*model.ThreadResponse, error) {
+	var threads []*model.ThreadResponse
+	for _, thread := range s.threads {
+		if unreadOnly && thread.UnreadReplies == 0 {
+			continue
+		}
+		threads = append(threads, cloneThreadResponse(thread, &model.ThreadResponse{}))
+	}
+	return threads, nil
+}
+
+// ThreadsSorted returns all threads, sorted by LastReplyAt
+func (s *MemStore) ThreadsSorted(unreadOnly, asc bool) ([]*model.ThreadResponse, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	threads, err := s.getThreads(unreadOnly)
+	if err != nil {
+		return nil, err
+	}
+	sort.Slice(threads, func(i, j int) bool {
+		if asc {
+			return threads[i].LastReplyAt < threads[j].LastReplyAt
+		}
+		return threads[i].LastReplyAt > threads[j].LastReplyAt
+	})
+	return threads, nil
+}
+
+func cloneThreadResponse(src *model.ThreadResponse, dst *model.ThreadResponse) *model.ThreadResponse {
```

Comment:
We usually take source and return a cloned struct. Taking dst in the parameter and returning it again seems slightly unidiomatic.

Reply:
Yeah, this is a bit iffy. I did it this way so that I could:

1. directly call this function in the parent's return statement, i.e. `return cloneThreadResponse(...)`
2. use inplace, e.g. `append(threads, cloneThreadResponse(thread, &model.ThreadResponse{}))`
3. pass the threadResponse object in the case where I need to update an existing object, i.e. in `Memstore.SetThread`